services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SemanticDetector.java

```diff
@@ -0,0 +1,438 @@
+   */
+  private static boolean isCorrectSchema(Object value, Schema schema) {
+    switch (schema.getType()) {
+      case STRING:
+        return value instanceof String;
+      case INT:
+        return value instanceof Integer;
+      case LONG:
+        return value instanceof Long;
+      case FLOAT:
+        return value instanceof Float;
+      case DOUBLE:
+        return value instanceof Double;
+      case BOOLEAN:
+        return value instanceof Boolean;
+      case ENUM:
+        return value instanceof String && schema.getEnumSymbols().contains(value);
+      case FIXED:
+        return (value instanceof byte[]) && ((byte[]) value).length == schema.getFixedSize();
+      case BYTES:
+        return value instanceof byte[];
+      case RECORD:
+        return value instanceof GenericRecord && ((GenericRecord) value).getSchema().equals(schema);
+      case ARRAY:
+        return value instanceof List
+            && ((List<?>) value).stream().allMatch(element -> isCorrectSchema(element, schema.getElementType()));
+      case MAP:
+        return value instanceof Map && ((Map<?, ?>) value).values()
+            .stream()
+            .allMatch(element -> isCorrectSchema(element, schema.getValueType()));
+      case NULL:
+        return value == null;
+      default:
+        throw new IllegalArgumentException("Unsupported schema type: " + schema.getType());
+    }
+  }
+
+  /**
+   * Validate the object with default value.
+   * There are two default values:
+   * <ol>
+   *   <li>Default value derived from schema as @code{@param defaultValue}} </li>
+   *   <li>Default value derived from the schema type of the object</li>
+   * </ol>
+   * @param object the object to validate
+   * @param currentSchema the schema of the object
+   * @param name the name of the object
+   * @param defaultValue the default value of the object
+   */
+  public static void compareObjectToDefaultValue(
+      Object object,
+      Schema currentSchema,
+      String name,
+      Object defaultValue) {
+    if (object == null || object.equals(defaultValue)) {
+      return;
+    }
+
+    switch (currentSchema.getType()) {
+      case INT:
+        if (!((object instanceof Integer) && (int) object == 0)) {
+          throw new VeniceProtocolException(
+              String.format(
+                  "Field %s: Integer value %s is not the default value 0 or %s",
+                  formatFieldName(name),
+                  object,
+                  defaultValue));
+        }
+        break;
+      case LONG:
+        if (!((object instanceof Long) && (long) object == 0L)) {
+          throw new VeniceProtocolException(
+              String.format(
+                  "Field %s: Long value %s is not the default value 0 or %s",
+                  formatFieldName(name),
+                  object,
+                  defaultValue));
+        }
+        break;
+      case FLOAT:
+        if (!((object instanceof Float) && (float) object == 0.0f)) {
```

Comment:
BTW, for double and float, I would recommend checking the difference between object and 0.0f or 0.0d, and confirm that the difference is smaller than a very small number like 0.00001.
It's not accurate to check object equals to 0.0